### PR TITLE
Pre commit flake8s

### PR DESCRIPTION
**Pull request by nstarman**
### Description

Now that `black` will be in, I'm adding a few flake8 checks. Most have no effect because we haven't made these mistakes. Only the F401 check made significant changes.
These checks shouldn't be burdensome for contributors.
When we eventually settle on all the checks we want to enforce we can switch to an exclude-list, not this verbose check-list.

Copied from https://github.com/astropy/astropy/issues/13735

- [x] E101 | indentation contains mixed spaces and tabs
- [x] E111 | indentation is not a multiple of four
- [x] E112 | expected an indented block
- [x] E113 | unexpected indentation
- [x] E201 | whitespace after ‘(’
- [x] E202 | whitespace before ‘)’
- [x] E301 | expected 1 blank line, found 0
- [x] E302 | expected 2 blank lines, found 0
- [x] E303 | too many blank lines (3)
- [x] E304 | blank lines found after function decorator
- [x] E305 | expected 2 blank lines after end of function or class
- [x] E306 | expected 1 blank line before a nested definition
- [x] E502 | the backslash is redundant between brackets
- [x] E722 | do not use bare except, specify exception instead
- [x] E901 | SyntaxError or IndentationError
- [x] E902 | IOError
- [x] E999
- [x] F401 : module imported but unused
- [x] F402 : import module from line N shadowed by loop variable
- [x] F404 : future import(s) name after other statements
- [x] F405 : name may be undefined, or defined from star imports: module
- [x] F406 : ‘from module import *’ only allowed at module level
- [x] F407 : an undefined __future__ feature name was imported
- [x] F501 : invalid % format literal
- [x] F502 : % format expected mapping but got sequence
- [x] F503 : % format expected sequence but got mapping
- [x] F504 : % format unused named arguments
- [x] F505 : % format missing named arguments
- [x] F506 : % format mixed positional and named arguments
- [x] F507 : % format mismatch of placeholder and argument count
- [x] F508 : % format with * specifier requires a sequence
- [x] F509 : % format with unsupported format character
- [x] F521 : .format(...) invalid format string
- [x] F522 : .format(...) unused named arguments
- [x] F523 : .format(...) unused positional arguments
- [x] F524 : .format(...) missing argument
- [x] F525 : .format(...) mixing automatic and manual numbering
- [x] F541 : f-string without any placeholders
- [x] F701 : a break statement outside of a while or for loop
- [x] F702 : a continue statement outside of a while or for loop
- [x] F703 : a continue statement in a finally block in a loop
- [x] F704 : a yield or yield from statement outside of a function
- [x] F706 : a return statement outside of a function/method
- [x] F707 : an except: block as not the last exception handler
- [x] F901 : raise NotImplemented should be raise NotImplementedError
- [x] W191 | indentation contains tabs
- [x] W291 | trailing whitespace
- [x] W292 | no newline at end of file
- [x] W293 | blank line contains whitespace
- [x] W391 | blank line at end of file
- [x] W605 | invalid escape sequence ‘x’
- [x] W606 | ‘async’ and ‘await’ are reserved keywords starting with Python 3.7

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Review by mcara — APPROVED** [review]
WCS part looks reasonable

**Review by WilliamJamieson — COMMENTED** [review]
The pre-commit changes look good to me. At first glance all the rest of the changes seem reasonable.

**Review by taldcroft — APPROVED** [review]
LGTM for table, time, io.ascii

**Review by dhomeier — APPROVED** [review]
`io` looks sensible to me, too, thanks!

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -115,7 +138,7 @@ repos:
           [
             "--count",
             "--select",
-            "E101,W191,E201,E202,W291,W292,W293,W391,E111,E112,E113,E30,E502,E722,E901,E902,E999,F822,F823",
+            "E101,E111,E112,E113,E201,E202,E30,E502,E722,E901,E902,E999,F401,F402,F404,F405,F406,F407,F501,F502,F503,F504,F505,F506,F507,F508,F509,F541,F701,F702,F703,F704,F706,F707,F822,F823,F901,W191,W291,W292,W293,W391",
```
Comment (eerovaher) [inline .pre-commit-config.yaml:141]: Would an exclude list be shorter?

File: astropy/units/format/console.py [hunk 2]
```diff
@@ -36,12 +36,12 @@ def _format_superscript(cls, number):
     @classmethod
     def _format_unit_list(cls, units):
         out = []
-        for base, power in units:
+        for bs, power in units:
```
Comment (eerovaher) [inline astropy/units/format/console.py:39]: Isn't it customary to avoid name conflicts by using trailing underscores?
```suggestion
        for base_, power in units:
```

**Review by eerovaher — COMMENTED** [review]
Overall I like these changes, but I do have two questions.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (nstarman) [reply .pre-commit-config.yaml:141]: 🤞 it soon will be. There's a fair number of https://pycodestyle.pycqa.org/en/latest/intro.html#error-codes

File: astropy/units/format/console.py (hunk 2 above)
Comment (nstarman) [reply astropy/units/format/console.py:39]: ✅ . Thanks!

File: astropy/coordinates/builtin_frames/__init__.py [hunk 3]
```diff
@@ -29,6 +29,10 @@
 from .baseradec import BaseRADecFrame
 from .cirs import CIRS
 from .ecliptic import *  # there are a lot of these so we don't list them all explicitly
```
Comment (eerovaher) [inline astropy/coordinates/builtin_frames/__init__.py:31]:
```suggestion
```

File: astropy/units/__init__.py [hunk 4]
```diff
@@ -16,6 +16,8 @@
 # this also makes it easier to understand where most time is spent
 # (e.g., using python -X importtime).
 
+from .core import set_enabled_units
+
 # isort: off
 from .core import *
```
Comment (eerovaher) [inline astropy/units/__init__.py:22]:
```suggestion
```

File: docs/conf.py [hunk 5]
```diff
@@ -59,6 +59,8 @@
     sys.exit(1)
 
 from sphinx_astropy.conf.v1 import *  # noqa: E402, F403
```
Comment (eerovaher) [inline docs/conf.py:61]:
```suggestion
```

File: astropy/units/__init__.py [hunk 6]
```diff
@@ -16,6 +16,8 @@
...
 from .quantity import *
```
Comment (eerovaher) [inline astropy/units/__init__.py:23]: Do we want to keep this line as it is?

**Review by eerovaher — CHANGES_REQUESTED** [review]
I found a few instances where an `import *` line was meant to be replaced with explicit listings, but the old `import *` line itself was not deleted.

**Comment by nstarman** [comment]
I left those there because I'm only importing a subset of the namespace. In another Issue (can't find it on the mobile app) we discuss replacing star imports with explicit imports, but that's TBD. This just explicitly imports functions and classes which cause F401 violations.

**Review by astrofrog — APPROVED** [review]
Looks good to me for timeseries and visualization

File: astropy/units/astrophys.py [hunk 7]
```diff
@@ -9,7 +9,7 @@
 from astropy.constants import si as _si
 
 from . import si
-from .core import UnitBase, binary_prefixes, def_unit, set_enabled_units, si_prefixes
+from .core import UnitBase, binary_prefixes, def_unit, set_enabled_units, si_prefixes  # noqa: F401
```
Comment (mhvk) [inline astropy/units/astrophys.py:12]: I actually think in this case we should just be able to remove the unused ones.

File: astropy/units/required_by_vounit.py [hunk 8]
```diff
@@ -15,8 +15,8 @@
 
 def _initialize_module():
     # Local imports to avoid polluting top-level namespace
-    from . import astrophys, cgs
-    from .core import _add_prefixes, def_unit
+    from . import astrophys, cgs  # noqa: F401
```
Comment (mhvk) [inline astropy/units/required_by_vounit.py:18]: Here too I think the unused imports can be removed (but I haven't tried...)

File: astropy/units/tests/test_structured.py [hunk 9]
```diff
@@ -10,7 +10,7 @@
 from numpy.testing import assert_array_equal
 
 from astropy import units as u
-from astropy.tests.helper import check_pickling_recovery, pickle_protocol
+from astropy.tests.helper import check_pickling_recovery, pickle_protocol  # noqa: F401
```
Comment (mhvk) [inline astropy/units/tests/test_structured.py:13]: Why is this necessary to mark it? Both are in fact used inside the module.

**Review by mhvk — COMMENTED** [review]
Just looked at units, mostly fine but a few questions...

Thanks for doing this cleanup!

File: astropy/units/tests/test_structured.py (hunk 9 above)
Comment (nstarman) [reply astropy/units/tests/test_structured.py:13]: Pickle protocol is a pytest fixture, so it's only referenced as a function argument. A flake8 plug-in to detect this would be nice!

**Comment by WilliamJamieson** [comment]
Looks like the merger of #13850 removed the usage of this [`import os`](https://github.com/astropy/astropy/blob/5c1b7160c4bbcda9a00840dabd7e89e58728769f/astropy/io/votable/tests/table_test.py#L6), which is what is causing the pre-commit.ci failure.

This is a good demonstration of why some of these stricter `flake8` rules should be enforced.

File: astropy/units/tests/test_structured.py (hunk 9 above)
Comment (mhvk) [reply astropy/units/tests/test_structured.py:13]: Ah, yes, makes sense.

File: astropy/units/__init__.py (hunk 6 above)
Comment (nstarman) [reply astropy/units/__init__.py:23]: ``.core`` is imported first, so it should be fine. It would eventually be nice to properly isort this module, but that's a task for another time :).

**Comment by nstarman** [comment]
@eerovaher, I switched to using a mixture of allow and block-lists.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (nstarman) [reply .pre-commit-config.yaml]: Done! It does simplify things considerably. Thanks for the suggestion.

**Review by mhvk — APPROVED** [review]
Units all OK now. Thanks!

File: astropy/coordinates/builtin_frames/__init__.py (hunk 3 above)
Comment (nstarman) [reply astropy/coordinates/builtin_frames/__init__.py:31]: Resolved for now by https://github.com/astropy/astropy/pull/13853#discussion_r1002588599.
In the future, this should be re-examined. There's #13785 for further discussion.

File: astropy/coordinates/builtin_frames/__init__.py [hunk 10]
```diff
@@ -29,6 +29,10 @@
...
+from .ecliptic import (
+    BarycentricMeanEcliptic, BarycentricTrueEcliptic, BaseEclipticFrame, CustomBarycentricEcliptic,
+    GeocentricMeanEcliptic, GeocentricTrueEcliptic, HeliocentricEclipticIAU76,
+    HeliocentricMeanEcliptic, HeliocentricTrueEcliptic)
```
Comment (eerovaher) [inline astropy/coordinates/builtin_frames/__init__.py:35]: The `from .ecliptic import *` line serves no purpose because the contents of `__all__` are imported explicitly anyways: 
https://github.com/astropy/astropy/blob/63aefe6955db7d82208638ce60f47266a1be270d/astropy/coordinates/builtin_frames/ecliptic.py#L11-L15

**Review by eerovaher — CHANGES_REQUESTED** [review]
I do think a line of code should be deleted, but the rest of the pull request looks very good.

**Comment by WilliamJamieson** [comment]
@nstarman the failing CI was fixed earlier. Please rebase.

File: astropy/coordinates/builtin_frames/__init__.py (hunk 10 above)
Comment (nstarman) [reply astropy/coordinates/builtin_frames/__init__.py]: ✅

**Review by eerovaher — APPROVED** [review]

File: docs/conf.py [hunk 11]
```diff
@@ -59,6 +59,8 @@
...
+from sphinx_astropy.conf.v1 import (
```
Comment (pllim) [inline docs/conf.py:62]: I must have missed this convo. tl;dr -- why the import star and then explicit import from the same place?

Comment (nstarman) [reply docs/conf.py:62]: https://github.com/astropy/astropy/pull/13853#discussion_r1002588599

Comment (nstarman) [reply docs/conf.py:62]: Basically, the explicit import is for F401, the star import is for everything else. We should be switching to explicit imports only, but that's a different convo — #13785

Comment (pllim) [reply docs/conf.py:62]: OK, thanks for the clarification!
